### ISSUE TITLE
Mocking ICosmosStore with Moq

### ISSUE THREAD

**Issue by reddy6ue** [Elfocrash/Cosmonaut#43]
I am facing the following problems with Moq. I am mocking an ICosmosStore to write unit tests in the service layer.

```
var fakeDataRepo = new Mock<ICosmosStore<FakeData>>();
var fakeData = Factory.CreateFakeData90;

activityRepo.Setup(p => 
						p.AddAsync(It.IsAny<FakeData>()))
						.Returns<FakeData>(a => 
								Task.FromResult(
									new CosmosResponse<FakeData>(fakeData, null)));
```

I get the error `Error	CS0854	An expression tree may not contain a call or invocation that uses optional arguments`

The root cause is that AddAsync has two other optional parameters RequestOptions and CancellationToken which trips up Moq. When I add the other two parameters, it obviously fails because I didn't setup the method that takes a single parameter. The problem is not with Cosmonaut per se, but it'll really help with a very popular mocking framework to also provide a version of AddAsync in the interface without the two optional parameter that just delegates to the method with the two optional parameter not being used.

**Comment by Elfocrash** [comment]
All you need to do is provide the default values for this call when you set it up. It's how Moq always worked and I have to do the same for my own tests as well.

Simply do:
```
activityRepo.Setup(p => p.AddAsync(It.IsAny<FakeData>(), null, default(CancellationToken)))
						.Returns<FakeData>(a => 
								Task.FromResult(
									new CosmosResponse<FakeData>(fakeData, null)));
```

**Comment by reddy6ue** [comment]
When I do this, I get a System.ArgumentException with the error `Invalid callback. Setup on method with 3 parameter(s) cannot invoke callback with different number of parameters (1)` error.

**Comment by Elfocrash** [comment]
That's weird. Can you change them with `It.IsAny`?

```
activityRepo.Setup(p => p.AddAsync(It.IsAny<FakeData>(), It.IsAny<RequestOptions>(), It.IsAny<CancellationToken>))
						.Returns<FakeData>(a => 
								Task.FromResult(
									new CosmosResponse<FakeData>(fakeData, null)));
```

**Comment by reddy6ue** [comment]
Tried that too. Didn't work. I think it has a problem with the fact that my method call sends only one parameter.

**Comment by Elfocrash** [comment]
That doesn't really make sense because calling a method with optional parameters means that the optional parameters are defaulted. They are still invoked though. Are you sure you are setting it up correctly? Did you try checking how i did it on my tests? I think it's the way you do `Returns`.

**Comment by reddy6ue** [comment]
I know it doesn't make sense. That's why I'm struggling with it. I changed my `AddAsync` methods to send the two additional parameters as `nul`l, and `default(CancellationToken)` and the error is still occurring. Baffling.

**Comment by Elfocrash** [comment]
Can you please paste the full test code so I can debug it locally?

**Comment by Elfocrash** [comment]
Btw, looking at the code you submitted initially, you are creating a fakeDataRepo but you are setting up an activityRepo. Is that intended?

**Comment by reddy6ue** [comment]
Meant to type in fakeDataRepo.

**Comment by Elfocrash** [comment]
Yeah, paste the test including any setup if possible please. I can’t reproduce it.

**Comment by theclai** [comment]
public class ActivityManagementServiceTest
    {
        private readonly Mock<IDocumentClient> _mockDocumentClient;
        private ActivityManagementService pActivityManagementService = null;

        public ActivityManagementServiceTest()
        {
            _mockDocumentClient = MockHelpers.GetMockDocumentClient();

            var activityRepo = new Mock<ICosmosStore<Activity>>();
            var boardRepo = new Mock<ICosmosStore<Board>>();
            var feedItemRepo = new Mock<ICosmosStore<FeedItem>>();
            var junaUserRepo = new Mock<ICosmosStore<JunaUser>>();
            var emojiRepo = new Mock<ICosmosStore<Emoji>>();
            var streamClient = new Mock<IStreamClient>();

            var fakeActivity = ActivityManagementServiceData.CreateActivityOne();

            var document = fakeActivity.ConvertObjectToDocument();
            var resourceResponse = document.ToResourceResponse(HttpStatusCode.OK);

            _mockDocumentClient.Setup(x => x.CreateDocumentAsync(UriFactory.CreateDocumentCollectionUri("databaseName", "dummies"),
                  document.ItIsSameDocument(), It.IsAny<RequestOptions>(), false, CancellationToken.None))
              .ReturnsAsync(resourceResponse);

            //var id = Guid.NewGuid().ToString();
            //var dummy = new Dummy
            //{
            //    Id = id,
            //    Name = "Nick"
            //};

            //var cosmoClient = new CosmonautClient(_mockDocumentClient.Object);
            //var activityRepo = new CosmosStore<Activity>(cosmoClient, "databaseName");

            activityRepo.Setup(p =>
                     p.AddAsync(It.IsAny<Activity>(),
                                 It.IsAny<RequestOptions>(),
                                 It.IsAny<CancellationToken>()))

                     .Returns<Activity>(a =>
                             Task.FromResult(
                                 new CosmosResponse<Activity>(resourceResponse)));

            //activityRepo.Verify();

            pActivityManagementService = new ActivityManagementService(
                activityRepo.Object,
                boardRepo.Object,
                feedItemRepo.Object,
                junaUserRepo.Object,
                emojiRepo.Object,
                streamClient.Object);
        }

        [Fact]
        public void StoreUniqueActivity_ValidInput_ReturnActivityObject()
        {
            var fakeActivity = ActivityManagementServiceData.CreateActivityOne();

            var storeActivity = pActivityManagementService.StoreUniqueActivityAsync(fakeActivity);

            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]

        public void StoreUniqueActivity_validInput_returnActivityObject()
        {
            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]

        public void DeleteActivity_validInput_ActivityDeleted()
        {
            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]
        public void LikeFeedItem_ValidInputParameter_Should_Be_LikedFeedItem()
        {
            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]
        public void UnLike_ValidInputParameter_Should_Be_UnLiked()
        {
            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]
        public void DislikeFeedItem_ValidInputParameter_Should_Be_DislikeFeedItem()
        {
            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]
        public void UndoDisLike_ValidInputParameter_Should_Be_UndoDisLiked()
        {
            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]
        public void ShareFeedItem_ValidInputParameter_Should_Be_ShareFeedItem()
        {
            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]
        public void UnshareFeedItem_ValidInputParameter_Should_Be_UnshareFeedItem()
        {
            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]
        public void PinFeedItem_ValidInputParameter_Should_Be_PinFeedItem()
        {
            Assert.True(true);
        }

        [Fact(Skip = "Not Implemented")]
        public void DeletePin_ValidInputParameter_Should_Be_DeletePin()
        {
            Assert.True(true);
        }
    }

> Yeah, paste the test including any setup if possible please. I can’t reproduce it.

**Comment by reddy6ue** [comment]
Here's the only thing I've done differently. I created a root entity class called `SharedCosmosEntity` which looks like this. 
```
	public interface ISharedCosmosEntity
	{
		string CosmosEntityName { get; set; }
	}
```
All these contortions in the code are to make sure that my collection name is the class name.
```
	public class SharedCosmosEntity : ISharedCosmosEntity
	{
		[JsonProperty("id")]
		public string Id { get; set; }

		// todo: This is not working. Need to figure out why
		[JsonProperty("CosmosEntityName")]
		public string CosmosEntityName
		{
			get => this.GetType().Name;
			set => value = this.GetType().Name;
		}
	}
```


And each entity looks like this. 

```	[SharedCosmosCollection("myColection", EntityName = "Activity")]
    public class Activity: SharedCosmosEntity
    {
         public string SomeProp {get; set;}
}
```

**Comment by reddy6ue** [comment]
Here's my test. 
```
    public class ActivityManagementServiceTest
    {
        private ActivityManagementService activityManagementService = null;

        public ActivityManagementServiceTest()
        {
            var activityRepo = new Mock<ICosmosStore<Activity>>();

            var fakeActivity = ActivityManagementServiceData.CreateActivityOne();

            var document = fakeActivity.ConvertObjectToDocument();
            var resourceResponse = document.ToResourceResponse(HttpStatusCode.OK);

            activityRepo.Setup(p =>
                     p.AddAsync(It.IsAny<Activity>(),
								 null, default(CancellationToken)))
                     .Returns<Activity>(a =>
                             Task.FromResult(
                                 new CosmosResponse<Activity>(fakeActivity, resourceResponse)));

            activityManagementService = new ActivityManagementService(activityRepo.Object);
        }

        [Fact]
        public void StoreUniqueActivity_ValidInput_ReturnActivityObject()
        {
            var fakeActivity = ActivityManagementServiceData.CreateActivityOne();

            var storeActivity = pActivityManagementService.StoreUniqueActivityAsync(fakeActivity);

            Assert.True(true);
        }
}
```

**Comment by Elfocrash** [comment]
You don not need to deal with the `CosmosEntityName` property. Cosmonaut will do that for you.
